cumulus/pallets/weight-reclaim-tx/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+)]
+pub struct StorageWeightReclaim<T, S>(pub S, core::marker::PhantomData<T>);
+
+impl<T, S> StorageWeightReclaim<T, S> {
+	/// Create a new `StorageWeightReclaim` instance.
+	pub fn new(s: S) -> Self {
+		Self(s, Default::default())
+	}
+}
+
+impl<T, S> From<S> for StorageWeightReclaim<T, S> {
+	fn from(s: S) -> Self {
+		Self(s, core::marker::PhantomData)
+	}
+}
+
+impl<T, S: core::fmt::Debug> core::fmt::Debug for StorageWeightReclaim<T, S> {
+	fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> Result<(), core::fmt::Error> {
+		#[cfg(feature = "std")]
+		let _ = write!(f, "StorageWeightReclaim<{:?}>", self.0);
+
+		Ok(())
+	}
+}
+
+// Make this extension "invisible" from the outside (ie metadata type information)
+impl<T, S: scale_info::StaticTypeInfo> scale_info::TypeInfo for StorageWeightReclaim<T, S> {
+	type Identity = S;
+	fn type_info() -> scale_info::Type {
+		S::type_info()
+	}
+}
+
+impl<T: Config + Send + Sync, S: TransactionExtensionBase> TransactionExtensionBase
+	for StorageWeightReclaim<T, S>
+{
+	const IDENTIFIER: &'static str = S::IDENTIFIER;
+	type Implicit = S::Implicit;
+
+	fn weight() -> Weight {
+		T::WeightInfo::storage_weight_reclaim().saturating_add(S::weight())
+	}
+}
+
+impl<T: Config + Send + Sync, S: TransactionExtension<T::RuntimeCall, Context>, Context>
+	TransactionExtension<T::RuntimeCall, Context> for StorageWeightReclaim<T, S>
+where
+	T::RuntimeCall: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>,
+{
+	type Val = (Option<u64>, S::Val);
+	type Pre = (Option<u64>, S::Pre);
+
+	fn validate(
+		&self,
+		origin: T::RuntimeOrigin,
+		call: &T::RuntimeCall,
+		info: &DispatchInfoOf<T::RuntimeCall>,
+		len: usize,
+		context: &mut Context,
+		self_implicit: Self::Implicit,
+		inherited_implication: &impl Encode,
+	) -> Result<(ValidTransaction, Self::Val, T::RuntimeOrigin), TransactionValidityError> {
+		// Trade-off: we could move it to `prepare` but better be accurate on reclaim than fast on
+		// `validate`
+		let proof_size = get_proof_size();
+
+		let res= self.0
+			.validate(origin, call, info, len, context, self_implicit, inherited_implication)?;
+
+		Ok((res.0, (proof_size, res.1), res.2))
+	}
+
+	fn prepare(
+		self,
+		val: Self::Val,
+		origin: &T::RuntimeOrigin,
+		call: &T::RuntimeCall,
+		info: &DispatchInfoOf<T::RuntimeCall>,
+		len: usize,
+		context: &Context,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		let (proof_size, inner_val) = val;
+		let inner_pre = self.0.prepare(inner_val, origin, call, info, len, context)?;
+
+		Ok((proof_size, inner_pre))
+	}
+
+	fn post_dispatch_details(
+		pre: Self::Pre,
+		info: &DispatchInfoOf<T::RuntimeCall>,
+		post_info: &PostDispatchInfoOf<T::RuntimeCall>,
+		len: usize,
+		result: &DispatchResult,
+		context: &Context,
+	) -> Result<Option<Weight>, TransactionValidityError> {
+		let (pre_dispatch_proof_size, inner_pre) = pre;
+
+		let inner_weight =
+			S::post_dispatch_details(inner_pre, info, post_info, len, result, context)?;
```

Comment:
I see, maybe the `post_dispatch_details` implementation for tuples should log a warning, or have a debug error.

Reply:
Implementing `post_dispatch_details` prevent us from returning the correct `pays_fee` if a transaction extension modified it.

It may be more accurate to implement `post_dispatch` I did at: https://github.com/paritytech/polkadot-sdk/pull/5234/commits/4d99a453e8b58c3dacb514fa11314857cb574f97